faculty/clients/experiment.py

```diff
@@ -293,3 +333,34 @@ def list_runs(
         return self._get(
             endpoint, ListExperimentRunsResponseSchema(), params=query_params
         )
+
+    def log_run_data(
+        self, project_id, run_id, metrics=None, params=None, tags=None
+    ):
+        """Update the data of a run.
+
+        Parameters
+        ----------
+        project_id : uuid.UUID
+        run_id : uuid.UUID
+        metrics : List[experiments.ExperimentRunMetric], optional
+            Each metric will be inserted.
+        params : List[experiments.ExperimentRunMetric], optional
+            Each param will be inserted and rejected if the given key is
+            present.
+        tags : List[experiments.ExperimentRunTag], optional
+            Each tag be upserted.
+
+        Returns
+        -------
+        None
+        """
+        endpoint = "/project/{}/run/{}/data".format(project_id, run_id)
+        payload = ExperimentRunDataSchema().dump(
+            {
+                "metrics": ExperimentRunMetricSchema(many=True).dump(metrics),
+                "params": ExperimentRunParamSchema(many=True).dump(params),
+                "tags": ExperimentRunTagSchema(many=True).dump(tags),
+            }
+        )
+        return self._patch_raw(endpoint, json=payload)
```

Comment:
I take it we're eventually going to interpret the response from the server to determine when e.g. metric updates have failed? Will this happen in a later PR?

Reply:
It's only the params that can cause a 409 and I was going to handle that specifically in the plugin. I believe that it would be correctly mapped to a `faculty.clients.base.HttpError.Conflict` as it stands now?

Reply:
> I was going to handle that specifically in the plugin.

That might be the right solution, however since it's usually the responsibility of a client to deserialise information in the response we might want to consider converting the 409 to a custom exception type with structured information inside in the client.

> I believe that it would be correctly mapped to a `faculty.clients.base.HttpError.Conflict` as it stands now?

Yeah, I think so.

Reply:
The server can also return a 404 if the run does not exist.

Reply:
Also, the server gives back a detailed message in case there is a parameter conflict. It feels like we can do a lot more here, as Andrew said.

Reply:
After chatting with Pascal we decided to implement a custom error for the Conflict case but not for the 404 as we don't in the other methods. This doesn't mean that we shouldn't do it and retrofit it to existing methods but it won't be part of this PR